### PR TITLE
password check, userName and profile pic on the navbar after login

### PR DESCRIPTION
**Pull request by MayaIw**
- Added checking for strong password
- Dissabled duplicate emails registration
- Added userName and profile pic to the navbar, after successful login.

### CODE REVIEW

**Comment by MayaIw** [comment]
@irisrifold@gmail.com

File: client/src/App.js [hunk 1]
```diff
@@ -6,36 +6,37 @@ import Login from "./pages/Login/Login";
...
-            <Route path="/upload" element={<UploadDataPage/>} />
-          </Routes>
-        </main>
-        <footer className={styles.footer}>
-          <p>&copy; 2024 My App</p>
-        </footer>
-      </div>
-    </BrowserRouter>
+    <AuthProvider>
+      <BrowserRouter>
```
Comment (irisri) [inline client/src/App.js]: move the BrowserRouter to the top

File: client/src/App.js [hunk 2]
```diff
@@ -6,36 +6,37 @@ import Login from "./pages/Login/Login";
...
+        <div className={styles.app}>
+          <header className={styles.appHeader}>
+            <img
+              src="/project-logo.png"
+              alt="Logo"
+              className={styles.appLogo}
+            />
+            <Navbar></Navbar>
```
Comment (irisri) [inline client/src/App.js]: ** Navbar => NavBar
`<NavBar />`

File: client/src/App.js [hunk 3]
```diff
@@ -6,36 +6,37 @@ import Login from "./pages/Login/Login";
...
+          <main className={styles.main}>
+            <Routes>
+              <Route path="/" element={<Home />} />
+              <Route path="/register" element={<Register />} />
+              <Route path="/login" element={<Login />} />
+              <Route path="/attraction" element={<AttractionOverviewPage />} />
+              <Route path="/upload" element={<UploadDataPage />} />
+            </Routes>
+          </main>
+          <footer className={styles.footer}>
```
Comment (irisri) [inline client/src/App.js:34]: can be moved to a seperated component

File: client/src/components/NavBar/NavBar.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { Link } from "react-router-dom";
+import { useContext } from "react";
+import { AuthContext } from "../../context/AuthContext";
+import styles from "../../styles/App.module.css";
+
+const Navbar = () => {
+  const { user } = useContext(AuthContext);
+  /*if (!user) {
+    return <div>Loading...</div>;
```
Comment (irisri) [inline client/src/components/NavBar/NavBar.js]: delete comment

File: client/src/components/NavBar/NavBar.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { Link } from "react-router-dom";
+import { useContext } from "react";
+import { AuthContext } from "../../context/AuthContext";
+import styles from "../../styles/App.module.css";
+
+const Navbar = () => {
```
Comment (irisri) [inline client/src/components/NavBar/NavBar.js]: export const Navbar

File: client/src/components/NavBar/NavBar.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        Home
+      </Link>
+      <div>
+        <img
+          src={user.profilePic}
+          alt="profile pic"
+          className={styles.appLogo}
+        />
+        <div>{user.userName}</div>
+      </div>
```
Comment (irisri) [inline client/src/components/NavBar/NavBar.js]:
```
{user.isLoggedIn ? 
     ( <div>
        <img
          src={user.profilePic}
          alt="profile pic"
          className={styles.appLogo}
        />
        <div>{user.userName}</div>
      </div>) : (
 <Link to="/login" className={styles.appLink}>
          Log In / Register
        </Link>
)
```

File: client/src/components/User/LoginForm.js [hunk 7]
```diff
@@ -1,10 +1,15 @@
-import { useState } from "react";
+import { useContext, useState } from "react";
 import { Link } from "react-router-dom";
+import { AuthContext } from "../../context/AuthContext";
 
 const LoginForm = () => {
+  /*const [loggedIn, setLoggedIn] = useState(false);
+  const [userName, setUserName] = useState("");*/
```
Comment (irisri) [inline client/src/components/User/LoginForm.js]: delete comment

File: client/src/components/User/LoginForm.js [hunk 8]
```diff
@@ -18,10 +23,11 @@ const LoginForm = () => {
     const json = await response.json(); //contains the userName and profilePic URL
```
Comment (irisri) [inline client/src/components/User/LoginForm.js]: please add try catch

File: client/src/components/User/LoginForm.js [hunk 9]
```diff
@@ -18,10 +23,11 @@ const LoginForm = () => {
...
-      setError(json.error);
+      setError("Incorrect username or password");
     } else {
       const name = json.userName;
-      // const profilePic = json.profilePic;
+      const profilePic = json.profilePic;
+      updateAuthState(name, profilePic);
       setEmail("");
       setPassword("");
       alert("Welcome " + name);
```
Comment (irisri) [inline client/src/components/User/LoginForm.js]: delete the alert

File: client/src/components/User/UserForm.js [hunk 10]
```diff
@@ -11,6 +11,23 @@ const UserForm = () => {
...
+      setError("Password must contain at least one capital letter");
+      return;
+    }
+
+    if (!/[!@#$%^&*(),.?":{}|<>~[\]\\/'`_+=-]/.test(password)) {
+      setError("Password must contain at least one spacial symbol");
+      return;
+    }
+
     const response = await fetch("http://localhost:5000/api/users", {
```
Comment (irisri) [inline client/src/components/User/UserForm.js:31]: try catch

File: client/src/context/AuthContext.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import { createContext, useState } from "react";
+
+// Create a context for the user data
+export const AuthContext = createContext();
+
+// AuthProvider component to wrap our app and provide user data to other components
+const AuthProvider = ({ children }) => {
```
Comment (irisri) [inline client/src/context/AuthContext.js:12]: delete un-nessecery context and providers

File: client/src/context/AuthContext.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  const [user, setUser] = useState({
+    isLoggedIn: false,
+    userName: "",
+    profilePic: "",
+  });
+
+  // update the state after a user is logged in
+  const updateAuthState = async (
+    userName,
+    profilePic = "https://www.pngitem.com/pimgs/m/391-3918613_personal-service-platform-person-icon-circle-png-transparent.png"
```
Comment (irisri) [inline client/src/context/AuthContext.js:17]: this should

File: client/src/context/AuthContext.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+// AuthProvider component to wrap our app and provide user data to other components
+const AuthProvider = ({ children }) => {
+  const [user, setUser] = useState({
+    isLoggedIn: false,
+    userName: "",
+    profilePic: "",
+  });
+
+  // update the state after a user is logged in
+  const updateAuthState = async (
```
Comment (irisri) [inline client/src/context/AuthContext.js]: should not put asunc in setState

File: client/src/context/AuthContext.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
+import { createContext, useState } from "react";
+
```
Comment (irisri) [inline client/src/context/AuthContext.js:7]: add 
`const defaultImage =  "https://www.pngitem.com/pimgs/m/391-3918613_personal-service-platform-person-icon-circle-png-transparent.png"`

File: client/src/components/User/LoginForm.js [hunk 15]
```diff
@@ -1,10 +1,15 @@
...
+import { AuthContext } from "../../context/AuthContext";
 
 const LoginForm = () => {
+  /*const [loggedIn, setLoggedIn] = useState(false);
+  const [userName, setUserName] = useState("");*/
   const [email, setEmail] = useState("");
   const [password, setPassword] = useState("");
+  //const [profilePicURL, setProfilePicURL] = useState("");
   const [error, setError] = useState(null);
+  const { updateAuthState } = useContext(AuthContext);
```
Comment (irisri) [inline client/src/components/User/LoginForm.js]: maybe move the login http request to contect

File: client/src/context/AuthContext.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+  ) => {
+    setUser({
+      isLoggedIn: true,
+      userName,
+      profilePic,
+    });
+  };
+
+  return (
+    <AuthContext.Provider value={{ user, updateAuthState }}>
```
Comment (irisri) [inline client/src/context/AuthContext.js]: delete the export of updateAuthState insted add a login function with http request

File: server/controllers/userController.js [hunk 17]
```diff
@@ -18,18 +18,25 @@ const getSingleUser = async (req, res) => {
...
-      email,
-      password: await bcrypt.hash(password, salt),
-      profilePicUrl,
-    });
-    res.status(200).json({ user });
-  } catch (err) {
-    res.status(400).json({ mssg: "error creating user", err });
+  const user_exists = await User.findOne({ email });
+  if (user_exists) {
+    res.status(400).json({ mssg: "user already exists" });
```
Comment (irisri) [inline server/controllers/userController.js:25]: return

File: server/controllers/userController.js [hunk 18]
```diff
@@ -18,18 +18,25 @@ const getSingleUser = async (req, res) => {
...
+  } else {
```
Comment (irisri) [inline server/controllers/userController.js]: you can delet the else

File: server/controllers/userController.js [hunk 19]
```diff
@@ -18,18 +18,25 @@ const getSingleUser = async (req, res) => {
...
+        const user = await User.create({
+          userName,
+          email,
+          password: await bcrypt.hash(password, salt),
+          profilePicURL,
+        });
+        res.status(200).json({ user });
+      }
+    } catch (err) {
+      res.status(400).json({ mssg: "error creating user", err });
```
Comment (irisri) [inline server/controllers/userController.js]: status 500
return err

File: server/controllers/userController.js [hunk 20]
```diff
@@ -48,7 +55,7 @@ const signIn = async (req, res) => {
       // check if the password matches
```
Comment (irisri) [inline server/controllers/userController.js:55]: line 51 statue code 401

File: server/controllers/userController.js [hunk 21]
```diff
@@ -48,7 +55,7 @@ const signIn = async (req, res) => {
...
       res
         .status(200)
-        .json({ userName: user.userName, profilePic: user.profilePicUrl });
+        .json({ userName: user.userName, profilePic: user.profilePicURL });
     } else {
       res.status(400).json({ mssg: "invalid password" });
```
Comment (irisri) [inline server/controllers/userController.js:60]: bad reques

**Review by irisri — APPROVED** [review]
Please add screenshots next time so I can see the changes without needing to checkout the branch

File: client/src/context/AuthContext.js [hunk 22]
```diff
@@ -0,0 +1,56 @@
+import { createContext, useState } from "react";
+import axiosInstance from "../services/api";
```
Comment (irisri) [inline client/src/context/AuthContext.js:2]: just call it axios

**Pull request merged**